src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+					}
+				}
+
+				async ValueTask<(bool, EndPoint?, ClusterMessages.MemberInfo[]?)> TryDiscover(int attempt) {
+					foreach (var channel in _grpcChannelsCache.Values.ToArray().OrderBy(_ => Guid.NewGuid())) {
+						try {
+							var clusterInfo = await gossipClient.GetAsync(channel, cancellationToken)
+								.ConfigureAwait(false);
+
+							if (clusterInfo.Members.Length == 0) {
+								continue;
+							}
+
+							var nodes = clusterInfo.Members.Where(IsCandidate).ToArray();
+
+							var node = nodes.OrderBy(node => node.State, comparer)
+								.ThenBy(_ => Guid.NewGuid())
+								.FirstOrDefault();
+
+							if (node is null) {
+								continue;
+							}
+
+							return (true, node.EndPoint, nodes);
+						} catch (Exception ex) {
+							_log.Log(attempt switch {
+									1 => LogLevel.Warning,
+									_ when attempt == _settings.ConnectivitySettings.MaxDiscoverAttempts => LogLevel
+										.Error,
+									_ => LogLevel.Debug
+								}, ex,
+								"Could not discover candidate. Attempts remaining: {discoverAttempts}",
+								_settings.ConnectivitySettings.MaxDiscoverAttempts - 1);
+						}
+					}
+
+					await Task.Delay(settings.ConnectivitySettings.DiscoveryInterval, cancellationToken)
+						.ConfigureAwait(false);
+
+					return (false, default, default);
+				}
+			}, cancellationToken);
+		}
+
+		private static bool IsCandidate(ClusterMessages.MemberInfo node) => node.IsAlive &&
+		                                                                    !NotAllowedStates.Contains(node.State);
+
+		public async Task<ChannelInfo> SelectChannel(CancellationToken cancellationToken = default) {
+			ThrowIfDisposed();
+
+			try {
+				return await _channelInfoChannel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+			} catch (Exception ex) when (ex.InnerException is DiscoveryException dex) {
+				throw dex;
+			}
+		}
+
+		public void Rediscover(EndPoint? leader) {
+			ThrowIfDisposed();
+
+			var message = leader is null
+				? (ControlMessage)ControlMessage.Discover.Instance
+				: new ControlMessage.SetEndPoint(leader);
+
+			_controlChannel.Writer.TryWrite(message);
+		}
+
+		private void ThrowIfDisposed() {
+			if (Interlocked.CompareExchange(ref _disposed, 0, 0) != 0) {
+				throw new ObjectDisposedException(GetType().ToString());
+			}
+		}
+
+		public async ValueTask DisposeAsync() {
+			if (Interlocked.Exchange(ref _disposed, 1) == 1) {
+				return;
+			}
+
+			_channelInfoChannel.Writer.TryComplete();
+			foreach (var channel in _grpcChannelsCache.Values) {
+				await channel.DisposeAsync().ConfigureAwait(false);
+			}
+		}
+
+		private async Task UseEndPoint(EndPoint endPoint) {
+			if (!_grpcChannelsCache.TryGetValue(endPoint, out var grpcChannel)) {
+				grpcChannel = ChannelFactory.CreateChannel(_settings, endPoint);
+				_grpcChannelsCache.Add(endPoint, grpcChannel);
+			}
+
+			if (!_serverCapabilitiesCache.TryGetValue(endPoint, out var serverCapabilities)) {
```

Comment:
is there a mechanism that requeries the caps in case of a server upgrade (or especially downgrade actually)? if not perhaps it'd be better not to cache them (?)

Reply:
I added a way to evict it